platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/analysis/PropertyAccessResolver.kt

```diff
@@ -73,8 +78,19 @@ class PropertyAccessResolverImpl(
     @Suppress("NestedBlockDepth")
     private
     fun AnalysisContext.doResolvePropertyAccessToObject(
-        propertyAccess: PropertyAccess
+        propertyAccess: PropertyAccess,
+        expectedType: DataTypeRef?
     ): ObjectOrigin? {
+        if (propertyAccess.receiver == null && expectedType != null && expectedType is DataTypeRef.Name) {
+            val dataType = schema.findTypeWithName(expectedType.fqName)
+            if (dataType is EnumClass) {
+                return dataType.entryNames.firstOrNull { it ==  propertyAccess.name}
+                    ?.let { ObjectOrigin.EnumConstantOrigin(dataType, propertyAccess) }
+            } else {
+                error("Unexpected expected type: $expectedType")
```

Comment:
This should not be an error. If the expected type is other than an enum class, we should just resolve the named reference normally.